packages/utils/src/lib/collect/implementation/persist.spec.ts

```diff
@@ -0,0 +1,160 @@
+
+    expect(() => readFileSync(outputPath + '.json')).toThrow(
+      'no such file or directory',
+    );
+    expect(() => readFileSync(outputPath + '.md')).toThrow(
+      'no such file or directory',
+    );
+  });
+
+  it('should log to console when format is stdout`', async () => {
+    await persistReport(dummyReport, {
+      ...dummyConfig,
+      persist: mockPersistConfig({ outputPath, format: ['stdout'] }),
+    });
+    /*
+    expect(console.log).toHaveBeenCalledTimes(configReportLogNum);
+    expect(
+      logs.find(msg => msg.match(/(erformance)*(11)*(EO)*(validators))),
+    ).toBeTruthy();
+    */
+    //
+    expect(() => readFileSync(outputPath + '.json')).toThrow(
+      'no such file or directory',
+    );
+    expect(() => readFileSync(outputPath + '.md')).toThrow(
+      'no such file or directory',
+    );
+  });
+
+  it('should persist json format`', async () => {
+    await persistReport(dummyReport, {
+      ...dummyConfig,
+      persist: mockPersistConfig({ outputPath, format: ['json'] }),
+    });
+    const jsonReport: Report = JSON.parse(
+      readFileSync(outputPath + '.json').toString(),
+    );
+    expect(jsonReport.plugins?.[0]?.meta.slug).toBe('plg-1');
+    expect(jsonReport.plugins?.[0]?.audits[0]?.slug).toBe('1a');
+    //
+    expect(console.log).toHaveBeenCalledTimes(0);
+    expect(() => readFileSync(outputPath + '.md')).toThrow(
+      'no such file or directory',
+    );
+  });
+
+  it('should persist md format`', async () => {
+    await persistReport(dummyReport, {
+      ...dummyConfig,
+      persist: mockPersistConfig({ outputPath, format: ['md'] }),
+    });
+    const mdReport = readFileSync(outputPath + '.md').toString();
+    expect(mdReport).toContain('# Code Pushup Report');
+    //
+    expect(console.log).toHaveBeenCalledTimes(0);
+    expect(() => readFileSync(outputPath + '.json')).toThrow(
+      'no such file or directory',
+    );
+  });
+
+  it('should persist all formats`', async () => {
+    await persistReport(dummyReport, {
+      ...dummyConfig,
+      persist: mockPersistConfig({
+        outputPath,
+        format: ['json', 'md', 'stdout'],
+      }),
+    });
+
+    //
+    const jsonReport: Report = JSON.parse(
+      readFileSync(outputPath + '.json').toString(),
+    );
+    expect(jsonReport.plugins?.[0]?.meta.slug).toBe('plg-1');
+    expect(jsonReport.plugins?.[0]?.audits[0]?.slug).toBe('1a');
+    //
+    const mdReport = readFileSync(outputPath + '.md').toString();
+    expect(mdReport).toContain('# Code Pushup Report');
+    //
+    /* expect(console.log).toHaveBeenCalledTimes(configReportLogNum);
+     expect(
+       logs.find(msg => msg.match(/(erformance)*(11)*(EO)*(validators))),
+     ).toBeTruthy(); */
+  });
+
+  it('should persist some formats`', async () => {
+    await persistReport(dummyReport, {
+      ...dummyConfig,
+      persist: mockPersistConfig({ outputPath, format: ['md', 'stdout'] }),
```

Comment:
I'm wondering whether we should allow omitting `json` output :thinking:  The `md` and `stdout` outputs are only for the user, but for uploading to the portal we need the `json` output.

It would be quite unpleasant to run a long `collect` command, only to find out when you run `upload` that you should've specified the `json` format. In general, I would say that when both `upload` config is specified and `persist.format` doesn't include `'json'`, that's almost certainly a mistake. We should either warn the user about it as soon as possible, or just create the `json` file regardless.

Reply:
> The md and stdout outputs are only for the user
The MD is also important for CI pipeline actions and the automated comments in PR's

Yes warning and also Ill see if I can just automatically create it in cases it makes sense

Reply:
So now it will get always generated